### PR TITLE
torch.compile: build graph for top-level TorchInGraph functions

### LINKED ISSUE #168284: torch.compile, when passes a single torch operation, does not put it into a graph
### 🐛 Describe the bug

Running these two pieces of code:

```
a = torch.compile(torch.exp)(x)
b = torch.compile(lambda x: torch.exp(x))(x)
```

behave differently. The first will just call the python operation, the second will capture it in a graph and go through the compile stack.

This is surprising because there is not reason, from a user perspective, for these to be different.
I might for example compile individual ops to opt into gemm autotuning, or for accuracy checking.

It is extra confusing if passing `fullgraph=True`, where we'd expect that either all pytorch operations are captured into a single graph, or we receive an error.

### Versions

nightly

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo @chenyang78

### PR DESCRIPTION
**Pull request by manigkrish**
Fixes #168284

Summary:
Previously, `torch.compile(torch.exp)(x)` would bypass Dynamo and run eager, while `torch.compile(lambda x: torch.exp(x))(x)` would go through Dynamo and build a graph. This is because `torch.exp` is a top-level function mapped to `TorchInGraphFunctionVariable`, but Dynamo did not wrap it with a frame when used directly as `torch.compile(torch.exp)`.

This PR makes top-level functions mapped to `TorchInGraphFunctionVariable` behave consistently with the lambda case by routing them through `wrap_inline`, so Dynamo can build a 1-node FX graph for them.

Changes:
- In `_TorchDynamoContext.__call__`, detect when the top-level `fn` is a `TorchInGraphFunctionVariable` target via `trace_rules.lookup(fn)`.
- When such a function is compiled, treat it like other “builtins without a frame” and call `external_utils.wrap_inline(fn)`, so Dynamo can trace the wrapper and emit a single FX node for the op.
- Add a regression test `SingleOpCompileTests.test_top_level_torch_exp_compiles_through_dynamo` in `test/dynamo/test_trace_rules.py` to assert that:
  - `torch.compile(lambda t: torch.exp(t))` and `torch.compile(torch.exp)` both go through Dynamo (FRAME_COUNTER increases), and Their outputs match.

This PR only addresses the first part of #168284 (handling top-level `TorchInGraphFunctionVariable` functions). The `fullgraph=True` behavior when no graph is traced will be handled in a separate PR as requested.

Test Plan:
- `pytest test/dynamo/test_trace_rules.py -vv` - Passed all test in local

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

### CODE REVIEW

File: test/dynamo/test_trace_rules.py [hunk 1]
```diff
@@ -527,6 +527,35 @@ def test_module_survive_skip_files(self):
         )
 
 
+class SingleOpCompileTests(torch._dynamo.test_case.TestCase):
+    def test_top_level_torch_exp_compiles_through_dynamo(self):
+        x = torch.randn(4)
+
+        # Sanity: lambda version should go through Dynamo
+        frame_count_before = torch._dynamo.convert_frame.FRAME_COUNTER
```
Comment (williamwen42) [inline test/dynamo/test_trace_rules.py]: We normally use the `torch._dynamo.testing.CompileCounter()` backend to check for this.

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -847,6 +847,35 @@ def aot_compile(example_inputs: tuple[tuple[Any, ...], dict[str, Any]]) -> Any:
...
+            rule = trace_rules.lookup(fn)
+        except Exception:
+            rule = None
+
+        if isinstance(rule, type):
+            torch_in_graph_cls = getattr(
+                trace_rules, "TorchInGraphFunctionVariable", None
+            )
+            if torch_in_graph_cls is not None and issubclass(rule, torch_in_graph_cls):
+                top_level_in_graph = True
```
Comment (williamwen42) [inline torch/_dynamo/eval_frame.py]: This code doesn't seem like it's actually doing anything?

**Review by williamwen42 — CHANGES_REQUESTED** [review]

**Review by williamwen42 — COMMENTED** [review]
approach seems fine to me but code seems to require some fixes

File: torch/_dynamo/eval_frame.py (hunk 2 above)
Comment (williamwen42) [reply torch/_dynamo/eval_frame.py]: Seems overly complicated, can we not just do (with proper imports)
```suggestion
        from .variables import TorchInGraphFunctionVariable
        top_level_in_graph = issubclass(trace_rules.lookup(fn), TorchInGraphFunctionVariable)
```

Comment (manigkrish) [reply torch/_dynamo/eval_frame.py]: I used your suggested `issubclass(trace_rules.lookup(fn)`, `TorchInGraphFunctionVariable)` and wrapped it in a local `rule = trace_rules.lookup(fn)` + `isinstance(rule, type)` to satisfy the PYREFLY type check (since lookup can be None). Behavior is unchanged, lint and tests are passing.

**Review by williamwen42 — COMMENTED** [review]
changes lgtm but want to quickly run this by @anijain2305 in case you think something could unexpectedly break

**Review by williamwen42 — APPROVED** [review]
